presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2706,24 +2709,47 @@ private boolean analyzeLimit(FetchFirst node)
             return node.isWithTies();
         }
 
-        private boolean analyzeLimit(Limit node)
+        private boolean analyzeLimit(Limit node, Scope scope)
         {
-            if (node.getLimit().equalsIgnoreCase("all")) {
-                analysis.setLimit(node, OptionalLong.empty());
+            OptionalLong rowCount;
+            if (node.getRowCount() instanceof AllRows) {
+                rowCount = OptionalLong.empty();
+            }
+            else if (node.getRowCount() instanceof LongLiteral) {
+                rowCount = OptionalLong.of(((LongLiteral) node.getRowCount()).getValue());
             }
             else {
-                long rowCount;
-                try {
-                    rowCount = Long.parseLong(node.getLimit());
+                checkState(node.getRowCount() instanceof Parameter, "unexpected LIMIT rowCount: " + node.getRowCount().getClass().getSimpleName());
+                if (analysis.isDescribe()) {
+                    analyzeExpression(node.getRowCount(), scope);
+                    analysis.addCoercion(node.getRowCount(), BIGINT, false);
+                    rowCount = OptionalLong.empty();
                 }
-                catch (NumberFormatException e) {
-                    throw semanticException(TYPE_MISMATCH, node, "Invalid LIMIT row count: %s", node.getLimit());
-                }
-                if (rowCount < 0) {
-                    throw semanticException(NUMERIC_VALUE_OUT_OF_RANGE, node, "LIMIT row count must be greater or equal to 0 (actual value: %s)", rowCount);
+                else {
+                    // validate parameter index
+                    analyzeExpression(node.getRowCount(), scope);
+                    Expression providedValue = analysis.getParameters().get(NodeRef.of(node.getRowCount()));
+                    try {
+                        rowCount = OptionalLong.of((long) ExpressionInterpreter.evaluateConstantExpression(
+                                providedValue,
+                                BIGINT,
+                                metadata,
+                                session,
+                                accessControl,
+                                analysis.getParameters()));
+                    }
+                    catch (VerifyException e) {
+                        throw semanticException(TYPE_MISMATCH, node, "Non constant parameter value for LIMIT: %s", providedValue);
```

Comment:
TYPE_MISMATCH is not an appropriate code for this error. Let's go with INVALID_ARGUMENTS